File: clients/java3/client/src/org/openamq/client/handler/ConnectionStartMethodHandler.java
package org.openamq.client.handler;

import org.apache.log4j.Logger;
import org.openamq.AMQException;
import org.openamq.client.state.AMQState;
import org.openamq.client.state.AMQStateManager;
import org.openamq.client.state.StateAwareMethodListener;
import org.openamq.client.protocol.AMQMethodEvent;
import org.openamq.client.protocol.AMQProtocolSession;
import org.openamq.framing.ConnectionStartBody;
import org.openamq.framing.ConnectionStartOkBody;
import org.openamq.framing.FieldTable;

import java.io.UnsupportedEncodingException;
import java.util.StringTokenizer;

public class ConnectionStartMethodHandler implements StateAwareMethodListener
{

    private static final Logger _log = Logger.getLogger(ConnectionStartMethodHandler.class);
    
    private static final ConnectionStartMethodHandler _instance = new ConnectionStartMethodHandler();
    
    public static ConnectionStartMethodHandler getInstance()
    {
        return _instance;
    }

    private ConnectionStartMethodHandler()
    {       
    }
    
    public void methodReceived(AMQStateManager stateManager, AMQMethodEvent evt) throws AMQException
    {
        ConnectionStartBody body = (ConnectionStartBody) evt.getMethod();
        
        try
        {
            if (body.mechanisms == null)
            {
                throw new AMQException("mechanism not specified in ConnectionStart method frame");
            }
            final String mechanisms = new String(body.mechanisms, "utf8");
            StringTokenizer tokenizer = new StringTokenizer(mechanisms, " ");
            String selectedMechanism = null;
            while (tokenizer.hasMoreTokens())
            {
                String mechanism = tokenizer.nextToken();
                if (mechanism.equals("PLAIN"))
                {
                    selectedMechanism = mechanism;
                    break;
                }
            }
    
            if (selectedMechanism == null)
            {
                throw new AMQException("No supported security mechanism found, passed: " + mechanisms);
            }
            
            FieldTable response = new FieldTable();
            final AMQProtocolSession ps = evt.getProtocolSession();
            response.put("login", ps.getUsername());
            response.put("password", ps.getPassword());

            if (body.locales == null)
            {
                throw new AMQException("Locales is not defined in Connection Start method");
            }
            final String locales = new String(body.locales, "utf8");
            tokenizer = new StringTokenizer(locales, " ");
            String selectedLocale = null;
            if (tokenizer.hasMoreTokens())
            {
                selectedLocale = tokenizer.nextToken();           
            }
            else
            {
                throw new AMQException("No locales sent from server, passed: " + locales);
            }        
                
            byte[] encodedResponse = response.getDataAsBytes();
            stateManager.changeState(AMQState.CONNECTION_NOT_TUNED);
            ps.writeFrame(ConnectionStartOkBody.createAMQFrame(evt.getChannelId(), selectedMechanism, encodedResponse,
                                                selectedLocale));
        }
        catch (UnsupportedEncodingException e)
        {
            throw new AMQException(_log, "Unable to decode data: " + e, e);
        }
    }

}

File: dotnet/xmsclient/Client/Handler/ConnectionStartMethodHandler.cs
using System;
using System.Text;
using log4net;
using OpenAMQ.Framing;
using OpenAMQ.XMS.Client.Protocol;
using OpenAMQ.XMS.Client.State;

namespace OpenAMQ.XMS.Client.Handler
{
    public class ConnectionStartMethodHandler : IStateAwareMethodListener
    {

        private static readonly ILog _log = LogManager.GetLogger(typeof(ConnectionStartMethodHandler));
        
        private static readonly ConnectionStartMethodHandler _instance = new ConnectionStartMethodHandler();
        
        public static ConnectionStartMethodHandler GetInstance()
        {
            return _instance;
        }

        private ConnectionStartMethodHandler()
        {       
        }
        
        public void MethodReceived(AMQStateManager stateManager, AMQMethodEvent evt)
        {
            ConnectionStartBody body = (ConnectionStartBody) evt.Method;
            
            try
            {
                if (body.Mechanisms == null)
                {
                    throw new AMQException("mechanism not specified in ConnectionStart method frame");
                }
                string allMechanisms = Encoding.ASCII.GetString(body.Mechanisms);
                string[] mechanisms = allMechanisms.Split(' ');
                string selectedMechanism = null;
                foreach (string mechanism in mechanisms)
                {
                    if (mechanism.Equals("PLAIN"))
                    {
                        selectedMechanism = mechanism;
                        break;
                    }
                }
        
                if (selectedMechanism == null)
                {
                    throw new AMQException("No supported security mechanism found, passed: " + mechanisms);
                }
                
                FieldTable response = new FieldTable();
                AMQProtocolSession ps = evt.ProtocolSession;
                response["login"] = ps.Username;
                response["password"] = ps.Password;

                if (body.Locales == null)
                {
                    throw new AMQException("Locales is not defined in Connection Start method");
                }
                string allLocales = Encoding.ASCII.GetString(body.Locales);
                string[] locales = allLocales.Split(new char[] { ' ' });
                string selectedLocale;
                if (locales != null && locales.Length > 0)
                {
                    selectedLocale = locales[0];
                }
                else
                {
                    throw new AMQException("No locales sent from server, passed: " + locales);
                }

                byte[] encodedResponse = response.GetDataAsBytes();
                stateManager.ChangeState(AMQState.CONNECTION_NOT_TUNED);
                FieldTable clientProperties = new FieldTable();
                clientProperties["product"] = "OpenAMQ JMS";
                clientProperties["version"] = "1.0";
                clientProperties["platform"] = GetFullSystemInfo();
                clientProperties["copyright"] = "(C) JP Morgan Chase 2006";
                ps.WriteFrame(ConnectionStartOkBody.CreateAMQFrame(evt.ChannelId, clientProperties, selectedMechanism,
                                                                   encodedResponse, selectedLocale));
            }
            catch (Exception e)
            {
                throw new AMQException(_log, "Unable to decode data: " + e, e);
            }
        }

        private string GetFullSystemInfo()
        {
            /*StringBuffer fullSystemInfo = new StringBuffer();
            fullSystemInfo.append(System.getProperty("java.runtime.name"));
            fullSystemInfo.append(", " + System.getProperty("java.runtime.version"));
            fullSystemInfo.append(", " + System.getProperty("java.vendor"));
            fullSystemInfo.append(", " + System.getProperty("os.arch"));
            fullSystemInfo.append(", " + System.getProperty("os.name"));
            fullSystemInfo.append(", " + System.getProperty("os.version"));
            fullSystemInfo.append(", " + System.getProperty("sun.os.patch.level"));*/
            // TODO: add in details here
            return ".NET 1.1 Client";
        }
    }
}
